interfaces/utils/path_patterns.go

```diff
@@ -0,0 +1,172 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package utils
+
+import (
+	"fmt"
+	"regexp"
+)
+
+type PathPattern struct {
+	pattern string
+	regex   *regexp.Regexp
+}
+
+const maxGroupDepth = 50
+
+type GlobFlags int
+
+const (
+	globDefault GlobFlags = 1 << iota
+	globNull
+)
+
+// createRegex converts the apparmor-like glob sequence into a regex. Loosely
+// using this as reference:
+// https://gitlab.com/apparmor/apparmor/-/blob/master/parser/parser_regex.c#L107
+func createRegex(pattern string, glob GlobFlags) (string, error) {
```

Comment:
we should document/explain somewhere that this set of functions does not work the same as the apparmor upstream when it comes to handling the `^` and `$` characters - we don't support them being in character classes here whereas the linked C implementation does

Reply:
or actually I guess is this a bug in our implementation? I'm mainly just referring to this handling in the C code:
```c
			/* these are special outside of character
			 * classes but not in them */
		case '^':
		case '$':
			if (incharclass) {
				pcre.append(1, *sptr);
			} else {
				pcre.append("\\");
				pcre.append(1, *sptr);
			}
			break;
```

since we don't have special cases for those characters when they are in a character class vs not in a character class, we seem to always quote these characters via QuotaMeta IIUC

Reply:
Yes, I had a little bug in there :-) It's now fixed, and I've added a unit test for it.